### DISCUSSION TITLE
Request for feedback: Contextual inventory availability

### DISCUSSION THREAD

**Discussion by coleatkinson** [Shopify/storefront-api-feedback#35] [RFC]
Traditionally Shopify's Online Store and Storefront API count inventory across all locations, and return a product as available so long as the total inventory number is > 1. This has been problematic when you are selling from multiple locations to multiple countries, as it does not take into account the customers context relative to your shipping profiles and locations.

For example:
- You have a Canada location which exclusively ships to Canada and a US location that ships exclusively to USA in your shipping profiles.
- If inventory for a product is 0 in Canada but 1+ in US, a Canadian buyer still sees the product as available and can purchase, sending the Canadian location into negative inventory unexpectedly.

We are fixing this in the 2021-10 API version, with fields `quantity_available`, `available_for_sale`, and `currently_not_in_stock` respecting the `inContext` declarative.

By using the [localization form](https://shopify.dev/themes/internationalization/multiple-currencies-languages#the-country-selector) to place your customers into the correct country context, the example above would now show the same product as out of stock to a Canadian buyer but in stock to a US buyer.

This will not yet flow through to checkout i.e. a buyer could still select USA on the storefront and then change to Canada at checkout to produce the same problem; something we are addressing in the near future. We'd love to hear any comments or feedback on this change, which you can test on the Unstable `2021-10` API version now.

**Comment by buzinas** [comment]
@coleatkinson Did this ever ship? I'm trying to achieve this exact use case: a client of mine has one warehouse in the UK an another in Germany. The first one ships only to the UK and the second one to European Union countries.

Using the `@inContext` query, I'm able to get prices in £ vs €, but the inventory always comes with the sum of all locations.

Is there anything specific I'm missing?

**Reply by duncan-fairley** [reply]
It's still working on `unstable` but not on other branches last I checked (last month)

**Reply by buzinas** [reply]
For me it doesn't work even on `unstable` 😞

**Reply by duncan-fairley** [reply]
@buzinas - it does require [this to be turned on](https://help.shopify.com/en/manual/markets/inventory_and_fulfillment) by a Shopify MSM and shipping rules that bind specific warehouses to specific geographies.

**Reply by duncan-fairley** [reply]
Can confirm that this is still only possible on the `unstable` branch

**Comment by davidhousedev** [comment]
Another use-case for this feature is perishable products. When shipping items that need to be frozen or temperature controlled in transit, it's not always feasible for the product to be shipped over great distances.

In case it wasn't clear, we have this need because we sell perishable products on Shopify :)

**Comment by duncan-fairley** [comment]
Yeah, it's a pretty important feature. It's sort of hard to imagine a large company adopting markets without it. As it stands we have several very large stores that are running `unstable` in production on critical queries (PDP, PLP) because their inventory mix varies significantly by warehouse / region and cross-border shipments are prohibitively expensive.

**Comment by lraffo** [comment]
This is critical for our setup as well. We have 1 shopify store and using Shopify Markets, we have 6 regional sites with each region being fulfilled by 1 of 3 warehouses.

As our VP of Digital Product, I am not comfortable with the risk around having our business rely on an unstable branch, when will this be moved to a stable branch?

**Comment by buzinas** [comment]
@shopify-admins Can anyone please give us an update on this feature? What's the release plan?

**Comment by coleatkinson** [comment]
Hi everyone, thanks for your thoughts and feedback here. We are reevaluating our approach to this feature so it can scale to handle larger product catalogs. What we found in the initial release was that when a store has a large catalog and/or ships to a large amount of countries, the inventory calculations simply take too long.

We are investing in making this happen this year, but I don't have a timeline I can share with high confidence right now. I will keep everyone on this thread updated.

**Reply by clarkejt** [reply]
Hey @coleatkinson has there been any update yet to the timeline?

**Reply by KarllosSouza** [reply]
hi @coleatkinson @shopify-admins 👋🏽

Is there any update on this matter?

**Reply by duncan-fairley** [reply]
@KarllosSouza - I believe this will be handled with [Fulfillable Inventory](https://help.shopify.com/en/manual/shipping/setting-up-and-managing-your-shipping/fulfillable-inventory) announced at Shopify Editions based on [this scenario description](https://help.shopify.com/en/manual/shipping/setting-up-and-managing-your-shipping/fulfillable-inventory#example-sell-only-within-configured-shipping-zones) (sell only within configured shipping zones) from the docs:

> There are two of Sally’s signature Coconut Bar Soap in stock in the United States location, but there are zero in stock in the Canada location. The remaining two soaps can be purchased only by customers located in the United States. For customers in Canada, the soap is displayed as out of stock on the product page and they can't purchase it. In this example, Sally’s Soaps is prioritizing following their shipping zone configurations instead of making every single sale.

At the moment I'm not seeing it enabled in stores that otherwise meet the criteria so it might still be in the process of rolling out.

**Reply by coleatkinson** [reply]
Hi @KarllosSouza - @duncan-fairley is exactly right.

The beta we initially launched to handle this problem has been replaced by the recently released Fulfillable inventory feature. As I mentioned in earlier posts on this thread, our initial approach to contextualizing inventory across different markets did not scale as we intended, and fulfillable inventory addresses these shortcomings 🎉

As far as the Storefront API is concerned, if you have the fulfillable inventory feature enabled in your admin and are correctly using in `@inContext` directive in your queries, inventory fields will automatically be contextualized on the Unstable release of the current API. This will be the behaviour on Stable as of the 2024-04-01 API release.

**Reply by duncan-fairley** [reply]
This is an exciting milestone. Congrats @coleatkinson, have been tracking this for years!
